src/SmoFacade/Listener.cs

```diff
@@ -133,16 +133,67 @@ private static string AgListenerName(string dataSource)
+        return new Server(connBuilder.ToString(), credentialName);
+      }
+      catch (Exception e)
+      {
+        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS", e);
+      }
+    }
+
+    /// <summary>
+    ///   Resolves "agReplicaInstanceName" to a FQDNS name
+    ///   However on Unix OS, "agReplicaInstanceName" should be a complete domain - "{sub domain}.{second level domain}.{...}.{top level domain}" (eg: "xyz.abc.def.com")
+    ///   Therefore, we use {second-level} -> {top-level} part of the domain from "agListenerDomain" and append it to "agReplicaInstanceName"
+    /// </summary>
+    /// <param name="agListenerDomain">The complete domain for the SQL server AG listener</param>
+    /// <param name="agReplicaInstanceName">The specific name of a replica instance within the AG</param>
+    private static string ResolveDnsAndGetHostName(string agListenerDomain, string agReplicaInstanceName)
+    {
+      // sometimes instances/listners have ports and named instances. Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      (string listenerDomain, string listenerPort) = SplitDomainPort(agListenerDomain);
+      (string instanceDomain, string instancePort) = SplitDomainPort(agReplicaInstanceName);
+      // we want to add back instancePort if it exists, otherwise the server port
+      string portToUse = instancePort ?? listenerPort;
+      try
+      {
+        return $"{Dns.GetHostEntry(instanceDomain).HostName}{portToUse}";
+      }
+      catch (System.Net.Sockets.SocketException e)
+      {
+        Console.WriteLine($"Unable to get dns for instance '{instanceDomain}'\n{e}");
+
+        // Re-try with full domain appended to the instanceDomain by stripping off domain from listner and appending to instance
+        // eg: if listner is "abc.def.ghi" we want to append ".def.ghi" to the instance
+        string[] domainFragments = listenerDomain.Split('.');
+        domainFragments[0] = "";
+        string instanceDomainFull = $"{instanceDomain}{string.Join(".", domainFragments)}";
+
+        Console.WriteLine($"Retrying with full domain '{instanceDomainFull}");
+        return $"{Dns.GetHostEntry(instanceDomainFull).HostName}{portToUse}";
+      }
+    }
+
+    private static (string domain, string port) SplitDomainPort(string domainAndPort)
+    {
+      string domain = domainAndPort;
+      string port = null;
+      if (domainAndPort.Contains(','))
+      {
+        string[] fragments = domainAndPort.Split(new char[] {','}, 2);
+        domain = fragments[0];
+        port = $",{fragments[1]}";
+      }
+      else if (domainAndPort.Contains('\\')) // handling named instances the same way as ports
+      {
+        string[] fragments = domainAndPort.Split(new char[] {'\\'}, 2);
+        domain = fragments[0];
+        port = $"\\{fragments[1]}";
+      }
+      return (domain, port);

```

Comment:
This method needs unit tests

```C#
var splitChar = domainAndPort.Contains(',') ? ',' : domainAndPort.Contains('\\') ? ' \\' : null;
if(splitChar != null)
{
  string[] fragments = domainAndPort.Split(splitChar, 2);
  domain = fragments[0];
  port = $"{splitChar}{fragments[1]}";
}

return (domain, port);

```

Reply:
^^ ah yes, much better way of doing it - thanks! 😊 

Yes agreed, there can be many edge cases here - will look at creating some unit tests

Was just implementing the PR suggestions from above. 
For next step, as mentioned before, I'm going to look at using this class instead of doing the fragmentations myself: 
[`UriBuilder`](https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder?view=netcore-2.0)

Reply:
For `UriBuilder`, looks like it expects a well formatted URI (with protocol), which we don't have.
Plus I can't see functions there to segment out the domain name Itself.
(for ports, we are also handling named instances - which is SQL Sever specific so it won't be able to handle those either)

reverting back to using normal string split functions.